explodomatica.1
.TH EXPLODOMATICA "1" "September 2011" "Explodomatica" "User Commands"
.SH NAME
explodomatica \- generate explosion sound effects 
.SH SYNOPSIS
.B explodomatica 
[\fIOPTION\fR]... [\fIFILE\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
An explosion sound effect is generated, and saved as mono 44100Hz
PCM data in the specified wav file.
.TP
\fB\-d\fR, \fB\-\-duration\fR
Specifies the approximate duration in seconds the explosion
should last.  Fractional seconds are permitted.
.TP
\fB\-\-input filename\fR
Allows a 44100Hz mono wav file to be used as input rather
than using generated white noise as the input.
.TP
\fB\-l\fR, \fB\-\-nlayers\fR
Specifies the number of sound layers which should be used
to create each sub-explosion within the explosion.
.TP
\fB\-\-noreverb\fR
Suppress the reverb effect
.TP
\fB\-p\fR, \fB\-\-preexplosions\fR
Specifies the number of "preexplosions" to generate.
You can think of the "preexplosions" as the "ka" part
of "ka-BOOM!"
.TP
\fB\-\-pre-delay\fR
Specifies the approximate duration in seconds of the
"ka" part of "ka-BOOM!".  Default is 0.25 seconds.
The value is somewhat randomized and this value is 
more like an upper bound with zero being the lower
bound.
.TP
\fB\-\-pre-lp-factor\fR
Specifies the factor to use with the low pass filter
on the pre-explosion.  Values closer to zero do more
(suppress the pre-explosion more) while values closer
to 1 do less.
.TP
\fB\-\-pre-lp-iters\fR
Specifies how many times to apply the low pass filter
to the pre-explosion.  Default is 1.
.TP
\fB\-s\fR, \fB\-\-speedfactor\fR
Specifies the factor by which to speed up or slow down
the final explosion sound.  Values greater than 1.0 speed
the sound up, values less than 1.0 slow the sound down.
The default is 0.45.
.SH EXAMPLES
.TP
explodomatica --duration 2 --preexplosions 0 --nlayers 3 test.wav
.SH SEE ALSO
<http://scameron.github.com/explodomatica>
.SH AUTHOR
Written by Stephen M. Cameron
.SH COPYRIGHT
Copyright \(co 2011 Stephen M. Cameron.
.br
License GPLv2+: GNU GPL version 2 or later
.br
<http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


explodomatica.c

/* 
    (C) Copyright 2011, Stephen M. Cameron.

    This file is part of explodomatica.

    explodomatica is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    explodomatica is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with explodomatica; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 */
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <malloc.h>
#include <math.h>
#include <stdlib.h>
#include <sys/time.h>
#include <getopt.h>
#include <limits.h>

#include <sndfile.h> /* libsndfile */

#include "explodomatica.h"

void usage(void)
{
	fprintf(stderr, "usage:\n");
	fprintf(stderr, "explodomatica [options] somefile.wav\n");
	fprintf(stderr, "caution: somefile.wav will be overwritten.\n");
	fprintf(stderr, "options:\n");
	fprintf(stderr, "  --duration n    Specifies duration of explosion in secs\n");
	fprintf(stderr, "                  Default value is %f secs\n",
			explodomatica_defaults.duration);
	fprintf(stderr, "  --nlayers n     Specifies number of sound layers to use\n");
	fprintf(stderr, "                  to build up each explosion.  Default is %d\n", explodomatica_defaults.nlayers);
	fprintf(stderr, "  --preexplosions n\n");
	fprintf(stderr, "                  Specifies number of 'pre-explostions' to generate\n");
	fprintf(stderr, "                  Default is %d\n", explodomatica_defaults.preexplosions);
	fprintf(stderr, "  --pre-delay n\n");
	fprintf(stderr, "                  Specifies approximate length of the 'ka' in 'ka-BOOM!'\n");
	fprintf(stderr, "                  (it is somewhat randomized)\n");
	fprintf(stderr, "                  Default is %f secs\n", explodomatica_defaults.preexplosion_delay);
	fprintf(stderr, "  --pre-lp-factor n\n");
	fprintf(stderr, "                  Specifies the impact of the low pass filter used\n");
	fprintf(stderr, "                  on the pre-explosion part of the sound.  values\n");
	fprintf(stderr, "                  closer to zero lower the cutoff frequency\n");
	fprintf(stderr, "                  while values close to one raise it.\n");
	fprintf(stderr, "                  Value should be between 0.2 and 0.9.\n");
	fprintf(stderr, "                  Default is %f\n", explodomatica_defaults.preexplosion_low_pass_factor);
	fprintf(stderr, "  --pre-lp-count n\n");
	fprintf(stderr, "                  Specifies the number of times the low pass filter used\n");
	fprintf(stderr, "                  on the pre-explosion part of the sound.  values\n");

	fprintf(stderr, "                  Default is %d\n", explodomatica_defaults.preexplosion_lp_iters);
	
	fprintf(stderr, "  --speedfactor n\n");
	fprintf(stderr, "                  Amount to speed up (or slow down) the final\n");
	fprintf(stderr, "                  explosion sound. Values greater than 1.0 speed\n");
	fprintf(stderr, "                  the sound up, values less than 1.0 slow it down\n");
	fprintf(stderr, "                  Default is %f\n", explodomatica_defaults.final_speed_factor);
	fprintf(stderr, "  --noreverb      Suppress the 'reverb' effect\n");
	fprintf(stderr, "  --input file    Use the given (44100Hz mono) wav file\n"
			"                  as input instead of generating white noise for input.\n");
	exit(1);
}

static void process_options(int argc, char *argv[], struct explosion_def *e)
{
	int option_index = 0;
	int c, n, ival;
	double dval;

	static struct option long_options[] = {
		{"duration", 1, 0, 0},
		{"nlayers", 1, 0, 1},
		{"preexplosions", 1, 0, 2},
		{"speedfactor", 1, 0, 3},
		{"pre-delay", 1, 0, 4},
		{"pre-lp-factor", 1, 0, 5},
		{"pre-lp-count", 1, 0, 6},
		{"noreverb", 0, 0, 7},
		{"input", 1, 0, 8},
		{0, 0, 0, 0}
	};

	while (1) {

		c = getopt_long(argc, argv, "d:l:p:s:",
			long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 0: /* duration */
			n = sscanf(optarg, "%lg", &dval);
			if (n != 1)
				usage();
			e->duration = dval;
			printf("duration = %g\n", dval);
			break;
		case 1: /* nlayers */
			n = sscanf(optarg, "%d", &ival);
			if (n != 1)
				usage();
			printf("nlayers = %d\n", ival);
			e->nlayers = ival;
			break;
		case 2: /* preexplosions */
			n = sscanf(optarg, "%d", &ival);
			if (n != 1)
				usage();
			printf("preexplosions = %d\n", ival);
			e->preexplosions = ival;
			break;
		case 3: /* speedfactor */
			n = sscanf(optarg, "%lg", &dval);
			if (n != 1)
				usage();
			e->final_speed_factor = dval;
			printf("speedfactor = %g\n", dval);
			break;
		case 4: /* pre-delay */
			n = sscanf(optarg, "%lg", &dval);
			if (n != 1)
				usage();
			e->preexplosion_delay = dval;
			printf("preexplosion_delay = %g\n", dval);
			break;
		case 5: /* pre-lp-factor */
			n = sscanf(optarg, "%lg", &dval);
			if (n != 1)
				usage();
			e->preexplosion_low_pass_factor = dval;
			printf("preexplosion_low_pass_factor = %g\n", dval);
			break;
		case 6: /* preexplosion_lp_iters */
			n = sscanf(optarg, "%d", &ival);
			if (n != 1)
				usage();
			printf("preexplosion low pass count = %d\n", ival);
			e->preexplosion_lp_iters = ival;
			break;
		case 7: /* noreverb */
			printf("noreverb selected\n");
			e->reverb = 0;
			break;

		case 8: /* input file */
			strncpy(e->input_file, optarg, PATH_MAX);
			printf("input file: '%s'\n", e->input_file);
			break;
			
		default:
			usage();
		}
	}
	if (optind < argc) {
		strcpy(e->save_filename, argv[optind]);
		printf("save filename is %s\n", e->save_filename);
	} else
		usage();
}

int main(int argc, char *argv[])
{
	struct timeval tv;
	struct explosion_def e;

	e = explodomatica_defaults;

	gettimeofday(&tv, NULL);
	srand(tv.tv_usec);

	if (argc < 2)
		usage();

	process_options(argc, argv, &e);
	explodomatica(&e);

	return 0;
}

